pkg/reconciler/timeout_handler.go

```diff
@@ -101,6 +119,47 @@ func GetTimeout(d *metav1.Duration) time.Duration {
 	return timeout
 }
 
+// GetTaskRunEndTime calculates the time at which a TaskRun will be considered
+// timed out.
+func GetTaskRunEndTime(tr *v1alpha1.TaskRun) time.Time {
+	return tr.Status.StartTime.Time.Add(GetTimeout(tr.Spec.Timeout))
+}
+
+// GetBackoff records a count of the times it has seen a given StatusKey object
+// and calculates an appropriate backoff deadline based on that count. Only one
+// deadline per StatusKey may be active at any moment. Requests for a new backoff
+// in the face of an existing one will be ignored and details of the existing
+// backoff will be returned instead.
+//
+// Returned values are: the count of backoffs performed for this StatusKey, the
+// backoff's deadline and a boolean indicating whether the returned backoff is
+// newly calculated.
+func (t *TimeoutSet) GetBackoff(runObj StatusKey) (uint, time.Time, bool) {
+	t.backoffsMut.Lock()
+	defer t.backoffsMut.Unlock()
+	b := t.backoffs[runObj.GetRunKey()]
+	if time.Now().Before(b.deadline) {
+		return b.count, b.deadline, false
+	}
+	b.count += 1
+	b.deadline = time.Now().Add(backoffDuration(b.count))
+	t.backoffs[runObj.GetRunKey()] = b
+	return b.count, b.deadline, true
+}
+
+func backoffDuration(count uint) time.Duration {
```

Comment:
oh interesting, i hadn't considered injecting my own `rand.Intn` and gave up testing this func because i couldn't think of a way to get deterministic results.  Cheers, I'll give this a shot!

Reply:
I've extracted the jitter func, set it to `rand.Intn()` by default and added testcases for this.